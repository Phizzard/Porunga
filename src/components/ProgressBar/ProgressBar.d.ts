import { Palette } from '../../types'
/**
 * ProgressBar properties
 */
interface ProgressBarProps {
  /** The value of the progress bar (as percentage) */
  value?: number
  /** The color of the progress bar */
  color?: Palette
  /** Whether to add text in the progress bar */
  showValue?: boolean
  /** Text to display if custom */
  customText?: string
  /** Sets text color */
  textColor?: string
  /** Optionally round the progress bar */
  rounded?: boolean
  /** Optional class name */
  className?: string
}
/**
 * Normalize the value
 *
 * @param value the progress bar value
 */
export declare function flatten(value: number): number
/**
 * Get a string representation of the percentage
 *
 * @param value the progress bar value
 */
export declare function getStringPercent(value: number): string
/**
 * The ProgressBar component allows the user to create a flexible progress bar for use
 * in everything from loading states to step counters. You can choose to show the current
 * percent, a custom value, or add an optional rounding for the borders of the progress bar.
 */
declare const ProgressBar: ({
  value,
  color,
  showValue,
  customText,
  textColor,
  rounded,
  className,
}: ProgressBarProps) => JSX.Element
export default ProgressBar
