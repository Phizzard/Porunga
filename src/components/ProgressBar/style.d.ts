/// <reference types="react" />
/// <reference types="@emotion/core" />
declare type ProgressBarProps = {
  color?: string
  text?: string
  textColor?: string
  rounded?: boolean
}
export declare const Container: import('@emotion/styled-base').StyledComponent<
  import('react').DetailedHTMLProps<
    import('react').HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  >,
  ProgressBarProps,
  object
>
export declare const StyledProgressBar: import('@emotion/styled-base').StyledComponent<
  import('react').DetailedHTMLProps<
    import('react').HTMLAttributes<HTMLSpanElement>,
    HTMLSpanElement
  >,
  ProgressBarProps,
  object
>
export {}
