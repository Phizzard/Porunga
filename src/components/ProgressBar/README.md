Default Progress Bar

```jsx
<ProgressBar value={25} />
```

Progress Bar Accent Color Variant

```jsx
<ProgressBar color="accent-1" value={50} />
```

Progress Bar With Percentage Text

```jsx
<ProgressBar value={75} showValue />
```

Progress Bar With Custom Text

```jsx
<ProgressBar value={25} customText="1/4" />
```

Progress Bar (Rounded)

```jsx
<ProgressBar value={80} rounded />
```
