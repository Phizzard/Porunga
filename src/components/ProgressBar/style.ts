import styled from '@emotion/styled'
import { SerializedStyles } from '@emotion/core'

type ProgressBarProps = {
  color?: string
  text?: string
  textColor?: string
  rounded?: SerializedStyles
}

export const Container = styled.div<ProgressBarProps>`
  border: none;
  position: relative;
  display: block;
  max-width: 100%;
  background: rgba(0, 0, 0, 0.1);
  font-family: Open Sans, sans-serif;
  ${p => p.rounded}
`

export const StyledProgressBar = styled.span<ProgressBarProps>`
  display: block;
  background-color: ${p => p.color};
  height: 100%;
  border-radius: inherit;
  transition: width .2s ease-in;
  line-height: 1;
  height: 2rem;
  transition: width .1s ease, background-color .1s ease;
  ${p => p.rounded}

  :after {
    content: '${p => p.text}';
    color: ${p => p.textColor};
    font-weight: bold;
    height: 100%;
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    align-items: center;
    padding-right: 0.4rem;
  }
`
