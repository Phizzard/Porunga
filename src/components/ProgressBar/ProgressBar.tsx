import React from 'react'
import { useTheme } from 'emotion-theming'

import { get } from '../../utils/get'
import { ThemeConfig, Palette } from '../../types'
import { Container, StyledProgressBar } from './style'

/**
 * ProgressBar properties
 */
interface ProgressBarProps {
  /** The value of the progress bar (as percentage) */
  value?: number
  /** The color of the progress bar */
  color?: Palette
  /** Whether to add text in the progress bar */
  showValue?: boolean
  /** Text to display if custom */
  customText?: string
  /** Sets text color */
  textColor?: string
  /** Optionally round the progress bar */
  rounded?: boolean
  /** Optional class name */
  className?: string
}

/**
 * Normalize the value
 *
 * @param value the progress bar value
 */
export function flatten(value: number) {
  return Math.floor(value)
}

/**
 * Get a string representation of the percentage
 *
 * @param value the progress bar value
 */
export function getStringPercent(value: number) {
  return `${value}%`
}

/**
 * The ProgressBar component allows the user to create a flexible progress bar for use
 * in everything from loading states to step counters. You can choose to show the current
 * percent, a custom value, or add an optional rounding for the borders of the progress bar.
 */
const ProgressBar = ({
  value = 0,
  color = 'brand',
  showValue = false,
  customText = '',
  textColor = '#FFFFFF',
  rounded = false,
  className,
}: ProgressBarProps) => {
  const theme: ThemeConfig = useTheme()
  const flatVal = flatten(value)
  const valueAsString = getStringPercent(flatVal)
  const width = { width: valueAsString }
  const textToUse = showValue ? valueAsString : customText

  const progressBarProps = {
    color: get(theme, 'root', 'colors', color),
    text: textToUse,
    textColor,
  }

  return (
    <Container
      className={className}
      aria-valuemin={0}
      aria-valuemax={100}
      aria-valuenow={flatVal}
      role="progressbar"
      rounded={rounded}
    >
      <StyledProgressBar
        style={width}
        {...progressBarProps}
      ></StyledProgressBar>
    </Container>
  )
}

export default ProgressBar
