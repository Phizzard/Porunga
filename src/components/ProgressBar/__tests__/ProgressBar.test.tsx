import React from 'react'
import renderer from 'react-test-renderer'

import ProgressBar, { flatten, getStringPercent } from '../ProgressBar'

describe('ProgressBar', () => {
  it('renders correctly', () => {
    const component = renderer.create(<ProgressBar />)

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('flatten', () => {
    it('should return a flat number', () => {
      const actual = flatten(55.4)
      const expected = 55

      expect(actual).toBe(expected)
    })
  })

  describe('getStringPercent', () => {
    it('should return a stringified percentage', () => {
      const actual = getStringPercent(80)
      const expected = '80%'

      expect(actual).toBe(expected)
    })
  })
})
