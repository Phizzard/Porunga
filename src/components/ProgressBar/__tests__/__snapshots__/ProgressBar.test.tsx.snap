// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar renders correctly 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={0}
  className="css-1swweuh"
  role="progressbar"
>
  <span
    className="css-d42j4p"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;
